src/server/transaction.cc

```diff
@@ -71,12 +71,61 @@ void RecordTxScheduleStats(const Transaction* tx) {
   }
 }
 
+void RecordTxScheduleFastStats(const Transaction* tx, bool was_ooo, bool was_inline) {
+  DCHECK_EQ(tx->GetUniqueShard(), 1);
+
+  auto* ss = ServerState::tlocal();
+  if (was_ooo) {
+    ss->stats.tx_type_cnt[was_inline ? ServerState::INLINE : ServerState::QUICK]++;
+  } else {
+    ss->stats.tx_type_cnt[ServerState::NORMAL]++;
+  }
+  ss->stats.tx_width_freq_arr[0]++;
+}
+
 }  // namespace
 
 IntentLock::Mode Transaction::LockMode() const {
   return cid_->IsReadOnly() ? IntentLock::SHARED : IntentLock::EXCLUSIVE;
 }
 
+uint32_t Transaction::PhasedBarrier::Start(uint32_t count) {
+  DCHECK_EQ(DEBUG_Get(), 0u);
+  uint32_t phase = phase_.load(memory_order_relaxed);
+  count_.store(count, memory_order_release);
+  return phase;
+}
+
+bool Transaction::PhasedBarrier::Active() const {
+  return count_.load(memory_order_acquire) > 0;
+}
+
+void Transaction::PhasedBarrier::Dec(Transaction* keep_alive) {
+  // Prevent transaction from being destroyed after atomic was decreased and Wait() unlocked,
+  // but before foreign thread finished notifying.
+  // ::boost::intrusive_ptr guard(keep_alive);
+  CHECK_GT(keep_alive->GetUseCount(), 1u);  // Why would we call it without use_count??
+
```

Comment:
@romange Do I understand correctly, that with a long tx queue, it might occur that our poll jobs finish _before_ we execute, so we're hanging on WaitForShardCallbacks() with use_count = 1 and once we pass over this barrier, we can drop ourselves?

Because it somewhat feels artifical to me that we separately have use count and run count, where run_count should clearly be a subset of use_count, i.e. use_count should be twice higher, representing any potential referrer that can get stuck (either in the shard queue or in the tx queue)

But I can also keep the guard because it's simpler 🙂

Reply:
yes, you are correct. by shard queue  you mean `fb2::FiberQueue` ?